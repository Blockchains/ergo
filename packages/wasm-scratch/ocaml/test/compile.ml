open Ergo_wasm.Compiler

let m = new Wasm_module.t

let expressions =
  let open Ergo_lib.Core in
  [ ImpExprConst (Ejnull)
  ; ImpExprConst (Ejbool false)
  ; ImpExprConst (Ejbool true)
  ; ImpExprConst (Ejnumber 3.14)
  ; ImpExprConst (Ejstring ['e'; 'r'; 'g'; 'o'])
  ; ImpExprConst (Ejstring [])
  ]

let () =
  List.iter (fun x ->
      ignore (m#func ~result:[I32Type] (expr x m))
    ) expressions;
  Wasm.Print.module_ stdout 72 m#return

