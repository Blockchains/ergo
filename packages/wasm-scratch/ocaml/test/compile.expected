(module
  (type $0 (func (param i32) (result i32)))
  (type $1 (func (result i32)))
  (type $2 (func (param i32 i32) (result i32)))
  (type $3 (func))
  (memory $0 1)
  (global $0 (mut i32) (i32.const 0))
  (func $0
    (type 0)
    (local i32)
    (i32.const 0)
    (local.set 1)
    (local.get 1)
  )
  (func $1
    (type 0)
    (local i32)
    (i32.const 4)
    (local.set 1)
    (local.get 1)
  )
  (func $2
    (type 0)
    (local i32)
    (i32.const 8)
    (local.set 1)
    (local.get 1)
  )
  (func $3
    (type 0)
    (local i32)
    (i32.const 12)
    (local.set 1)
    (local.get 1)
  )
  (func $4
    (type 0)
    (local i32)
    (i32.const 24)
    (local.set 1)
    (local.get 1)
  )
  (func $5
    (type 0)
    (local i32)
    (i32.const 36)
    (local.set 1)
    (local.get 1)
  )
  (func $6
    (type 0)
    (local.get 0)
    (i32.load)
    (i32.const 1)
    (i32.le_u)
    (if (type 1) (then (i32.const 4)) (else (i32.const 8)))
  )
  (func $7
    (type 0)
    (local i32)
    (i32.const 8)
    (call 6)
    (local.set 1)
    (local.get 1)
  )
  (func $8
    (type 0)
    (local i32)
    (i32.const 4)
    (call 6)
    (local.set 1)
    (local.get 1)
  )
  (func $9
    (type 0)
    (local i32)
    (i32.const 0)
    (call 6)
    (local.set 1)
    (local.get 1)
  )
  (func $10
    (type 2)
    (local.get 0)
    (i32.load)
    (i32.const 1)
    (i32.le_u)
    (local.get 1)
    (i32.load)
    (i32.const 1)
    (i32.le_u)
    (i32.and)
    (if (type 1) (then (i32.const 4)) (else (i32.const 8)))
  )
  (func $11
    (type 0)
    (local i32)
    (i32.const 8)
    (i32.const 44)
    (call 10)
    (local.set 1)
    (local.get 1)
  )
  (func $12
    (type 2)
    (local.get 0)
    (i32.load)
    (i32.const 1)
    (i32.le_u)
    (local.get 1)
    (i32.load)
    (i32.const 1)
    (i32.le_u)
    (i32.or)
    (if (type 1) (then (i32.const 4)) (else (i32.const 8)))
  )
  (func $13
    (type 0)
    (local i32)
    (i32.const 4)
    (i32.const 8)
    (call 12)
    (local.set 1)
    (local.get 1)
  )
  (func $14 (type 3) (i32.const 56) (global.set 0))
  (export "memory" (memory 0))
  (export "alloc_p" (global 0))
  (export "f0" (func 0))
  (export "f1" (func 1))
  (export "f2" (func 2))
  (export "f3" (func 3))
  (export "f4" (func 4))
  (export "f5" (func 5))
  (export "f6" (func 7))
  (export "f7" (func 8))
  (export "f8" (func 9))
  (export "f9" (func 11))
  (export "f10" (func 13))
  (start 14)
  (data
    0
    (offset (i32.const 0))
    "\00\00\00\00\02\00\00\00\01\00\00\00\03\00\00\00"
    "\1f\85\eb\51\b8\1e\09\40\04\00\00\00\04\00\00\00"
    "\65\72\67\6f\04\00\00\00\00\00\00\00\03\00\00\00"
    "\00\00\00\00\00\00\f0\3f"
  )
)
